library/Vanilla/BodyFormatValidator.php

```diff
@@ -58,12 +58,46 @@ private function validateRich($value, $field, $row = []) {
             $value = new Invalid("%s is not valid rich text.");
         } else {
             // Re-encode the value to escape unicode values.
+            $this->stripUselessEmbedData($value);
             $value = json_encode($value);
         }
 
         return $value;
     }
 
+    /**
+     * There is certain embed data from the rich editor that we want to strip out. This includes
+     *
+     * - Malformed partially formed operations (dataPromise).
+     * - Nested embed data.
+     *
+     * @param array[] $operations The quill operations to loop through.
+     */
+    private function stripUselessEmbedData(array &$operations) {
```

Comment:
This method should probably have a name reflecting that it's specific to the rich format. This class could ultimately be used for any format and "strip useless embed data" might be confusing if you weren't familiar with the method, already. The name could probably be more generic too (e.g. `cleanRichData`).